winml/lib/Api/StringData.cpp

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "impl/StringData.h"
+
+namespace _winml {
+
+string_data::string_data(size_t size) :
+    buffer_(size) {}
+
+std::shared_ptr<_winml::idata> string_data::create(size_t size) {
+  return std::make_shared<string_data>(size);
+}
+
+size_t string_data::num_elements() {
+  return buffer_.size();
+}
+
+size_t string_data::size_in_bytes() {
+  WINML_THROW_HR(E_UNEXPECTED);
+}
+
+size_t string_data::num_buffers() {
+  return 1;
+}
+
+bool string_data::flush() {
+  // Vacuously true
+  return true;
+}
+
+std::vector<wss::IBuffer>& string_data::buffers() {
+  WINML_THROW_HR(E_UNEXPECTED);
+}
+
+gsl::span<byte> string_data::buffer(bool /*should_sync_buffer*/) {
+  return gsl::span<byte>(reinterpret_cast<byte*>(buffer_.data()), buffer_.size());
```

Comment:
This is a span of the string objects, not of string data. Is this expected?

Reply:
(And I think the size of the span is off, `vector<string>.size()` will give the number of strings in the vector, not its size in bytes)

Reply:
yeah, its onnxruntimevalue for strings returns data a pile of pointers to std::Strings... we return the same pile, and there is some glue in tensorbase to copy it in correctly.

Reply:
Huh, it's very counterintuitive

Reply:
Without further refactoring in tensorbase for string types this needs to stay the same.
I do envision a future change that will expose the underlying std::vector, and obviate the need for this weird tranformation into a byte buffer of std:strings.

Reply:
The reasoning for this is that we want to perform as few copies as possible for the tensor data.
We get back string tensors from ORT engine in a very weird format. All of the strings are collapsed into a single buffer, and a list of delimiters are added. We dont want to unnecessarily copy them around as that will be expensive, so we create string_views from them. Ultimately a copy is needed into the final backing tensor buffer. Therefore we need to convert an array of string_views into a vector of strings.

This was added during layering to quickly get things working on a new ort api stack.

To make this work in a more polymorphic way, another data type would needed to have been created for IValue's implementation as a StringTensor, and then QIs could have been done to avoid forcing strings down the void* buffer path.

On the Tensor<T> a better api can be introduced to avoid the string_view* but, i think that this api is actually fairly simple to understand. It is the ORT interaction before it that complicates matters.

It is more involved to make this change at this time.